### ISSUE TITLE
validate the path with PathSeparatedPrefix?

### ISSUE THREAD

**Issue by qicz** [envoyproxy/gateway#1044]
current gateway supports config match path with `PathSeparatedPrefix`, but does not validate in the gateway, when config path with invalid value like end with `/`, the gateway and envoy recv the dynamic route will occur error.

HTTPRoute config

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: backendxx
spec:
  parentRefs:
    - name: eg
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: backend
          port: 3000
          weight: 1
      matches:
        - path:
            type: PathPrefix
            value: /backend/
```

gateway log

```bash
2023-02-15 13:29:15 2023-02-15T05:29:15.921Z    INFO    cache/logrwrapper.go:29 Got a new request, response_nonce 2, nodeID envoy-default-eg-64656661-7db8b55db6-gttln, node_version v1.26.0    {"runner": "xds-server"}
2023-02-15 13:29:15 2023-02-15T05:29:15.921Z    INFO    cache/logrwrapper.go:29 handling v3 xDS resource request, response_nonce 2, nodeID envoy-default-eg-64656661-7db8b55db6-gttln, node_version v1.26.0, resource_names_subscribe [], resource_names_unsubscribe [], type_url type.googleapis.com/envoy.config.route.v3.RouteConfiguration, errorCode 13, errorMessage Proto constraint validation failed (RouteConfigurationValidationError.VirtualHosts[0]: embedded message failed validation | caused by VirtualHostValidationError.Routes[0]: embedded message failed validation | caused by RouteValidationError.Match: embedded message failed validation | caused by RouteMatchValidationError.PathSeparatedPrefix: value does not match regex pattern "^[^?#]+[^?#/]$"): name: "default-eg-http"
2023-02-15 13:29:15 virtual_hosts {
2023-02-15 13:29:15   name: "default-eg-http"
2023-02-15 13:29:15   domains: "*"
2023-02-15 13:29:15   routes {
2023-02-15 13:29:15     match {
2023-02-15 13:29:15       path_separated_prefix: "/backend/"
2023-02-15 13:29:15     }
2023-02-15 13:29:15     route {
2023-02-15 13:29:15       cluster: "default-backendxx-rule-0-match-0-*"
2023-02-15 13:29:15     }
2023-02-15 13:29:15   }
2023-02-15 13:29:15 }
2023-02-15 13:29:15     {"runner": "xds-server"}
```

envoy log

```bash
2023-02-15 13:29:15 [2023-02-15 05:29:15.920][1][warning][config] [source/common/config/new_delta_subscription_state.cc:288] delta config for type.googleapis.com/envoy.config.route.v3.RouteConfiguration rejected: Proto constraint validation failed (RouteConfigurationValidationError.VirtualHosts[0]: embedded message failed validation | caused by VirtualHostValidationError.Routes[0]: embedded message failed validation | caused by RouteValidationError.Match: embedded message failed validation | caused by RouteMatchValidationError.PathSeparatedPrefix: value does not match regex pattern "^[^?#]+[^?#/]$"): name: "default-eg-http"
2023-02-15 13:29:15 virtual_hosts {
2023-02-15 13:29:15   name: "default-eg-http"
2023-02-15 13:29:15   domains: "*"
2023-02-15 13:29:15   routes {
2023-02-15 13:29:15     match {
2023-02-15 13:29:15       path_separated_prefix: "/backend/"
2023-02-15 13:29:15     }
2023-02-15 13:29:15     route {
2023-02-15 13:29:15       cluster: "default-backendxx-rule-0-match-0-*"
2023-02-15 13:29:15     }
2023-02-15 13:29:15   }
2023-02-15 13:29:15 }
2023-02-15 13:29:15 
2023-02-15 13:29:15 [2023-02-15 05:29:15.920][1][warning][config] [source/common/config/grpc_subscription_impl.cc:128] gRPC config for type.googleapis.com/envoy.config.route.v3.RouteConfiguration rejected: Proto constraint validation failed (RouteConfigurationValidationError.VirtualHosts[0]: embedded message failed validation | caused by VirtualHostValidationError.Routes[0]: embedded message failed validation | caused by RouteValidationError.Match: embedded message failed validation | caused by RouteMatchValidationError.PathSeparatedPrefix: value does not match regex pattern "^[^?#]+[^?#/]$"): name: "default-eg-http"
2023-02-15 13:29:15 virtual_hosts {
2023-02-15 13:29:15   name: "default-eg-http"
2023-02-15 13:29:15   domains: "*"
2023-02-15 13:29:15   routes {
2023-02-15 13:29:15     match {
2023-02-15 13:29:15       path_separated_prefix: "/backend/"
2023-02-15 13:29:15     }
2023-02-15 13:29:15     route {
2023-02-15 13:29:15       cluster: "default-backendxx-rule-0-match-0-*"
2023-02-15 13:29:15     }
2023-02-15 13:29:15   }
2023-02-15 13:29:15 }
2023-02-15 13:29:15
```

but the route is accepted

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"gateway.networking.k8s.io/v1beta1","kind":"HTTPRoute","metadata":{"annotations":{},"name":"backendxx","namespace":"default"},"spec":{"parentRefs":[{"name":"eg"}],"rules":[{"backendRefs":[{"group":"","kind":"Service","name":"backend","port":3000,"weight":1}],"matches":[{"path":{"type":"PathPrefix","value":"/backend/"}}]}]}}
  creationTimestamp: "2023-02-15T05:28:38Z"
  generation: 2
  name: backendxx
  namespace: default
  resourceVersion: "209201"
  uid: abffbeb8-e808-46ca-80ad-0dad38802d65
spec:
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
  - backendRefs:
    - group: ""
      kind: Service
      name: backend
      port: 3000
      weight: 1
    matches:
    - path:
        type: PathPrefix
        value: /backend/
status:
  parents:
  - conditions:
    - lastTransitionTime: "2023-02-15T05:29:15Z"
      message: Route is accepted
      observedGeneration: 2
      reason: Accepted
      status: "True"
      type: Accepted
    controllerName: gateway.envoyproxy.io/gatewayclass-controller
    parentRef:
      group: gateway.networking.k8s.io
      kind: Gateway
      name: eg
```

I think that we should validate the PathSeparatedPrefix with an invalid path value and reject the route.

**Comment by arkodg** [comment]
@qicz can you please attach a config that causes the issue ?

**Comment by qicz** [comment]
> @qicz can you please attach a config that causes the issue ?

sorry, I am busy with others. the config is appended.

**Comment by qicz** [comment]
other idea, when the path.value end with '/' use the `RouteMatch_Prefix` like follow:

```golang
			if *pathMatch.Prefix == "/" {
				outMatch.PathSpecifier = &routev3.RouteMatch_Prefix{
					Prefix: "/",
				}
			} else if strings.HasSuffix(*pathMatch.Prefix, "/") {
				outMatch.PathSpecifier = &routev3.RouteMatch_Prefix{
					Prefix: *pathMatch.Prefix,
				}
			} else {
				outMatch.PathSpecifier = &routev3.RouteMatch_PathSeparatedPrefix{
					PathSeparatedPrefix: *pathMatch.Prefix,
				}
			}
```

**Comment by arkodg** [comment]
awesome thanks for root causing the issue, the fix LGTM
@skriss is something similar done in contour ?

adding a `help-wanted` label, hoping someone from the community can incorporate the above patch and add test cases

**Comment by qicz** [comment]
another case @arkodg 

```golang
type RouteMatch_PathSeparatedPrefix struct {
	// If specified, the route is a path-separated prefix rule meaning that the
	// ``:path`` header (without the query string) must either exactly match the
	// ``path_separated_prefix`` or have it as a prefix, followed by ``/``
	//
	// For example, ``/api/dev`` would match
	// ``/api/dev``, ``/api/dev/``, ``/api/dev/v1``, and ``/api/dev?param=true``
	// but would not match ``/api/developer``
	//
	// Expect the value to not contain ``?`` or ``#`` and not to end in ``/``
	PathSeparatedPrefix string `protobuf:"bytes,14,opt,name=path_separated_prefix,json=pathSeparatedPrefix,proto3,oneof"`
}
```

the `PathSeparatedPrefix` followed by ``/``. when the httproute config with filter urlrewrite like follow:

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: backend
spec:
  parentRefs:
    - name: eg
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: backend
          port: 3000
          weight: 1
      filters:
        - type: URLRewrite
          urlRewrite:
            path:
              replacePrefixMatch: /
              type: ReplacePrefixMatch
      matches:
        - path:
            type: PathPrefix
            value: /om/test
```

the upstream recv double `/`

![image](https://user-images.githubusercontent.com/2174082/219278148-3d8cd923-1bca-4745-bfde-4d5d0e721aa8.png)

**Comment by qicz** [comment]
> another case @arkodg
> 
> ```go
> type RouteMatch_PathSeparatedPrefix struct {
> 	// If specified, the route is a path-separated prefix rule meaning that the
> 	// ``:path`` header (without the query string) must either exactly match the
> 	// ``path_separated_prefix`` or have it as a prefix, followed by ``/``
> 	//
> 	// For example, ``/api/dev`` would match
> 	// ``/api/dev``, ``/api/dev/``, ``/api/dev/v1``, and ``/api/dev?param=true``
> 	// but would not match ``/api/developer``
> 	//
> 	// Expect the value to not contain ``?`` or ``#`` and not to end in ``/``
> 	PathSeparatedPrefix string `protobuf:"bytes,14,opt,name=path_separated_prefix,json=pathSeparatedPrefix,proto3,oneof"`
> }
> ```
> 
> when the httproute config with urlrewrite like follows:
> 
> ```yaml
> apiVersion: gateway.networking.k8s.io/v1beta1
> kind: HTTPRoute
> metadata:
>   name: backend
> spec:
>   parentRefs:
>     - name: eg
>   rules:
>     - backendRefs:
>         - group: ""
>           kind: Service
>           name: backend
>           port: 3000
>           weight: 1
>       filters:
>         - type: URLRewrite
>           urlRewrite:
>             path:
>               replacePrefixMatch: /
>               type: ReplacePrefixMatch
>       matches:
>         - path:
>             type: PathPrefix
>             value: /om/test
> ```
> 
> the upstream recv double `/`
> 
> ![image](https://user-images.githubusercontent.com/2174082/219278148-3d8cd923-1bca-4745-bfde-4d5d0e721aa8.png)
> 
> envoy doc
> 
> ![image](https://user-images.githubusercontent.com/2174082/219282696-055201f0-ffd1-4dc0-97ab-892ec84ed0b1.png)

I configured the `envoy.yaml` in envoy like following the case exists the same. maybe this is a normal state?

<img width="624" alt="image" src="https://user-images.githubusercontent.com/2174082/219526284-4b914681-0d39-45c5-904d-8616eb120c0d.png">

**Comment by kflynn** [comment]
Hey @qicz, this looks like excellent work! Is there something we did that stopped you from opening a PR? It feels like you literally did everything to just PR your fix, we'd love to be able to give you a shoutout for this... 🙂

**Comment by qicz** [comment]
fixed by #1125 ，close this issue
